xbmc/filesystem/PluginDirectory.cpp

```diff
@@ -476,14 +476,28 @@ bool CPluginDirectory::WaitOnScriptResult(const std::string &scriptPath, int scr
 {
   bool cancelled = false;
 
-  if (!m_fetchComplete.WaitMSec(20))
+  // CPluginDirectory::GetDirectory can be called from the main and other threads.
+  // If called form the main thread, we need to bring up the BusyDialog in order to
+  // keep the render loop alive
+  if (g_application.IsCurrentThread())
+  {
+    if (!m_fetchComplete.WaitMSec(20))
+    {
+      CScriptObserver scriptObs(scriptId, m_fetchComplete);
+      if (!CGUIDialogBusy::WaitOnEvent(m_fetchComplete, 200))
+      {
+        cancelled = true;
+      }
+      scriptObs.Abort();
+    }
+  }
+  else
   {
-    CScriptObserver scriptObs(scriptId, m_fetchComplete);
-    if (!CGUIDialogBusy::WaitOnEvent(m_fetchComplete, 200))
+    // kill the script if it does not return within 10 seconds
+    if (!m_fetchComplete.WaitMSec(10000))
```

Comment:
Will we get an obvious log message when the timeout is hit?
Seems for debugging it would be useful to distinguish timeouts (that may have eventually succeeded) from a clear error response which never would have succeeded.